addons/account/models/account_invoice.py

```diff
@@ -384,13 +399,34 @@ def get_view_id(xid, name):
                     view_id = get_view_id('invoice_form', 'account.invoice.form').id
         return super(AccountInvoice, self).fields_view_get(view_id=view_id, view_type=view_type, toolbar=toolbar, submenu=submenu)
 
+    @api.model
+    def invoice_report_post(self, report):
+        if len(self) > 1:
+            return report
+        if not self.sent and self.state in ('open', 'paid'):
+            attachment_name = 'INV%s.pdf' % self.number.replace('/', '')
+            attachment_id = self.env['ir.attachment'].create({
+                'name': attachment_name,
+                'datas': base64.encodestring(report),
+                'datas_fname': attachment_name,
+                'res_model': 'account.invoice',
+                'res_id': self.id,
+            })
```

Comment:
I would have put the attachment creation inside the report engine, triggered only with a key in the context (report_name)